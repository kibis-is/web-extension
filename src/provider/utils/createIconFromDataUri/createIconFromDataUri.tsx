import { Icon, IconProps } from '@chakra-ui/react';
import { sanitize } from 'dompurify';
import React, { ReactElement } from 'react';

/**
 * Creates a Chakra UI icon from a data URI svg string, base64 encoded or not.
 * @param {string} dataUri - the svg data URI to parse.
 * @param {IconProps} props - [optional] icon props to pass to the parsed icon.
 * @returns {ReactElement | null} the SVG data URI parsed as a Chakra UI icon or null if it failed to parse the data URI.
 * @see {@link https://chakra-ui.com/docs/components/icon#creating-your-custom-icons}
 */
export default function createIconFromDataUri(dataUri: string, props?: IconProps): ReactElement | null {
  let data: string | undefined;
  let document: Document;
  let parser: DOMParser;
  let svg: string;
  let viewBox: string | undefined;

  if (dataUri.indexOf('data:') !== 0) {
    return null;
  }

  if (dataUri.indexOf(',') < 0) {
    return null;
  }

  data = dataUri.split(',').pop();

  if (!data) {
    return null;
  }

  const [mimeType, encoding] = dataUri.substring('data:'.length, dataUri.indexOf(',')).split(';');

  if (mimeType !== 'image/svg+xml') {
    return null;
  }

  if (encoding === 'base64') {
    data = window.atob(data);
  }

  parser = new DOMParser();
  svg = decodeURIComponent(data);
  document = parser.parseFromString(sanitize(svg, { USE_PROFILES: { svg: true, svgFilters: true } }), mimeType); // sanitize for any nastiness
  viewBox = document.getElementsByTagName('svg').item(0)?.getAttribute('viewBox') || undefined;

  return (
    <Icon viewBox={viewBox} {...props}>
      {Array.from(document.getElementsByTagName('path')).map((value, index) => (
        <path d={value.getAttribute('d') || undefined} fill="currentColor" key={`path-${index}`} />
      ))}
    </Icon>
  );
}
