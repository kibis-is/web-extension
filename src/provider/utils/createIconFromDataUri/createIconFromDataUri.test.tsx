import { IconProps } from '@chakra-ui/react';
import { ReactNode } from 'react';

// utils
import createIconFromDataUri from './createIconFromDataUri';

interface ITestParams {
  error: string;
  input: string;
}

describe('createIconFromDataUri()', () => {
  const props: IconProps = { h: 3, w: 3 };

  it.each([
    {
      error: 'is in the incorrect scheme incorrect scheme',
      input: 'unknown',
    },
    {
      error: 'no data found',
      input: 'data:image/svg+xml',
    },
    {
      error: 'no data found',
      input: 'data:image/svg+xml;base64,',
    },
    {
      error: 'not an svg image',
      input: 'data:,Hello%2C%20World%21',
    },
  ])(`should return null if the data uri $error`, ({ input }: ITestParams) => {
    expect(createIconFromDataUri(input, props)).toBeNull();
  });

  it('should return a an icon from a base64 encoded string', () => {
    // Arrange
    const dataUri: string = `data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgMjAwIDIwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KICA8cGF0aCBkPSJNIDEwMCwgMTAwIG0gLTc1LCAwIGEgNzUsNzUgMCAxLDAgMTUwLDAgYSA3NSw3NSAwIDEsMCAtMTUwLDAiIC8+Cjwvc3ZnPgo=`;

    // Act
    const result: ReactNode | null = createIconFromDataUri(dataUri, props);

    // Assert
    expect(result).toMatchSnapshot();
  });

  it('should return a an icon from a string', () => {
    // Arrange
    const dataUri: string = `data:image/svg+xml,%3Csvg%20viewBox%3D%220%200%20200%20200%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%0A%20%20%3Cpath%20d%3D%22M%20100%2C%20100%20m%20-75%2C%200%20a%2075%2C75%200%201%2C0%20150%2C0%20a%2075%2C75%200%201%2C0%20-150%2C0%22%20%2F%3E%0A%3C%2Fsvg%3E`;

    // Act
    const result: ReactNode | null = createIconFromDataUri(dataUri, props);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
