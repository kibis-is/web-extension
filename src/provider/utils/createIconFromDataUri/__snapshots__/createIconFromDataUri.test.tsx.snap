// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIconFromDataUri() should return a an icon from a base64 encoded string 1`] = `
<Icon
  h={3}
  viewBox="0 0 200 200"
  w={3}
>
  <path
    d="M 100, 100 m -75, 0 a 75,75 0 1,0 150,0 a 75,75 0 1,0 -150,0"
    fill="currentColor"
  />
</Icon>
`;

exports[`createIconFromDataUri() should return a an icon from a string 1`] = `
<Icon
  h={3}
  viewBox="0 0 200 200"
  w={3}
>
  <path
    d="M 100, 100 m -75, 0 a 75,75 0 1,0 150,0 a 75,75 0 1,0 -150,0"
    fill="currentColor"
  />
</Icon>
`;
